yasmeen-alharbi/TruffleTalk
# Task/21 be registration and login

## Description
This PR introduces 3 new endpoints for registration, logging in, and logging out via Bearer token authentication. It also has a testing endpoint that will be removed in the future when more authed endpoints are introduced.

Laravel Fortify is used to handle user registration, as well as allow us additional user management like password reset and email verification that can be used in the future if desired.

Laravel Sanctum is used to authenticate all endpoints in the API that need to be authed using a token. If the user is not properly authenticated (logged in), then the endpoints will return 401 unauthorized.

All new endpoints live under `api.php`:
No auth needed:
- POST `/api/register` -> creates a new user if given valid input
- POST `/api/auth/token` -> generates a token if valid login credentials are passed (all subsequent requests while 'logged in' need this token)
Sanctum authed:
- DELETE `/api/auth/token` -> deletes all tokens for a given user with the token provided (acts as a logout)
- GET `/api/test` -> temporary endpoint that requires a valid token to be accessed


## Replication/Testing
Once the Laravel API is running using Sail, then:

Update `.env` with the new key/value pairs from the example env
- namely `SANCTUM_STATEFUL_DOMAINS` and `SESSION_DOMAIN`

Run migrations:
- `sail artisan migrate`

Update dependencies:
- `sail composer install`
You can check which routes are available in the API using `sail artisan route:list`. It should look something like this:
![image](https://github.com/yasmeen-alharbi/TruffleTalk/assets/93738138/c18082e5-9c98-4a2a-a592-074693d0af25)

You should now be ready to test using Postman!

The two collections below have test request bodies filled in and the necessary Bearer token setting to happen automatically.
[TruffleTalk - Bearer Token.postman_collection.json](https://github.com/yasmeen-alharbi/TruffleTalk/files/13301732/TruffleTalk.-.Bearer.Token.postman_collection.json)
[TruffleTalk - Unauthed.postman_collection.json](https://github.com/yasmeen-alharbi/TruffleTalk/files/13301733/TruffleTalk.-.Unauthed.postman_collection.json)

What you will need to set up manually are some variables at the environment level so:
- create a new environment, mine is called `localhost`
- create a `url` variable and set it to `http://localhost`
- create an empty `authToken` variable, which will get automatically set in the POST `/api/auth/token` request upon succes

If you run into any problems, let me know. Happy reviewing!
